commit 4727e7e4258ecee07ec83c0923b5f51ff3b1719a (refs/changes/19/819/2)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-10-31T11:41:12-07:00 (2 years, 11 months ago)
    
    DOCKER-959 Unable to pull from registry when registry response sends multiple Docker-Distribution-Api-Version headers

diff --git a/CHANGES.md b/CHANGES.md
index 00fe6b0..e1e5eea 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,11 @@
 
 ## not yet released
 
+## 3.2.4
+
+- DOCKER-959 Unable to pull from registry when registry response sends multiple
+  Docker-Distribution-Api-Version headers.
+
 ## 3.2.3
 
 - IMGAPI-596 Move from restify-clients fork to restify-clients@1.4.0 with
diff --git a/lib/registry-client-v1.js b/lib/registry-client-v1.js
index 7a07eae..a6c0b7c 100644
--- a/lib/registry-client-v1.js
+++ b/lib/registry-client-v1.js
@@ -770,7 +770,7 @@ RegistryClientV1.prototype.listRepoImgs = function listRepoImgs(cb) {
     }, function _afterListRepoImgs(err, req, res, repoImgs) {
         if (err) {
             cb(_sanitizeErr(err, res, res,
-                self.repo.localName + ' repo not found'));
+                self.repo.localName + ' v1 repo not found'));
         } else {
             cb(null, repoImgs, res);
         }
diff --git a/lib/registry-client-v2.js b/lib/registry-client-v2.js
index 6ffa9c9..fc22c1a 100644
--- a/lib/registry-client-v2.js
+++ b/lib/registry-client-v2.js
@@ -1190,7 +1190,18 @@ RegistryClientV2.prototype.supportsV2 = function supportsV2(cb) {
         if (res && (res.statusCode === 200 || res.statusCode === 401)) {
             var header = res.headers['docker-distribution-api-version'];
             if (header) {
-                var versions = header.split(/\s+/g);
+                /*
+                 * Space- or comma-separated. The latter occurs if there are
+                 * two separate headers, e.g.:
+                 *      $ curl -i https://registry.example.com/v2/
+                 *      HTTP/1.1 200 OK
+                 *      ...
+                 *      Docker-Distribution-Api-Version: registry/2.0
+                 *      ...
+                 *      Docker-Distribution-Api-Version: \
+                 */
+                // JSSTYLED
+                var versions = header.split(/[\s,]+/g);
                 if (versions.indexOf('registry/2.0') !== -1) {
                     return cb(null, true);
                 }
diff --git a/package.json b/package.json
index fc4db8c..a14d077 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
     "name": "docker-registry-client",
-    "version": "3.2.3",
+    "version": "3.2.4",
     "description": "node.js client for the Docker Registry API",
     "author": "Joyent (joyent.com)",
     "main": "./lib/index.js",
diff --git a/test/v2.quayio.test.js b/test/v2.quayio.test.js
index 1d4df71..4394459 100644
--- a/test/v2.quayio.test.js
+++ b/test/v2.quayio.test.js
@@ -20,7 +20,9 @@ var drc = require('..');
 var log = require('./lib/log');
 
 var REPO = 'quay.io/coreos/etcd';
-var TAG = 'latest';
+// Note: Not using TAG='latest' as a workaround for
+// <https://github.com/joyent/node-docker-registry-client/issues/12>.
+var TAG = 'v2.0.0';
 
 // Trent will report these to "support@quay.io" and jzelinskie on #quay IRC
 var SKIP_QUAY_IO_BUGLETS = true;
@@ -76,7 +78,8 @@ test('v2 quay.io', function (tt) {
             t.ifErr(err);
             t.ok(tags);
             t.equal(tags.name, repo.remoteName);
-            t.ok(tags.tags.indexOf(TAG) !== -1, 'no "'+TAG+'" tag');
+            t.ok(tags.tags.indexOf(TAG) !== -1,
+                'tag "'+TAG+'" in listTags:' + JSON.stringify(tags));
             t.end();
         });
     });
